### ISSUE TITLE
avoid_types_on_closure_parameters false positive with implicit-dynamic

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/linter#3330]
**Describe the issue**
Some SDK functions type their error callback parameter as `Function` to accept both `Function(Object)` and `Function(Object, StackTrace)`:

```dart
Future future;
future.then(..., onError: <this callback>);
```

As such, writing the following will trigger implicit-dynamic:

```dart
Future future;
future.then(..., onError: (err, stack) {
  // err and stack are both dynamic
});
```

The solution should be to explicitly write the type:

```dart
Future future;
future.then(..., onError: (Object err, StackTrace stack) {
  // fixed implicit-dynamic
});
```

but doing this now causes `avoid_types_on_closure_parameters` on both parameters

**Expected behavior**

I would expect avoid_types_on_closure_parameters to detect that Object/StackTrace are types more specific than `dynamic`, and therefore allow it.

**Additional context**

Possibly related to https://github.com/dart-lang/linter/issues/695 and https://github.com/dart-lang/linter/issues/2131

**Comment by rrousselGit** [comment]
Assuming the special casing is based on `Function` not `Future.then(onError:)` then I think that's fine

Because ultimately folks may write utilities around such as:

```dart
void doSomething(Function fn) {
  Stream().handleError(fn);
}

// will warn
doSomething((Object err, StackTrace stack) {

});
```

**Comment by a14n** [comment]
In my opinion the issue is more general and arises at several places (assignments, arguments, declarations...) where a `dynamic`, `Object`, `Function` is expected. For instance in the following case I don't think the lint should triggers:

```dart
  dynamic f1;
  f1 = (int a) {}; // LINT
  Function f2;
  f2 = (int a) {}; // LINT
  Object f3;
  f3 = (int a) {}; // LINT
  var l = [
    (int a) => 1, // LINT
  ];
```

**Comment by srawlins** [comment]
Fixed with https://github.com/dart-lang/sdk/commit/f07e042eaf90c1f68e346068ba8e62f18946dc24
